packages/server/src/threads/automation.ts

```diff
@@ -330,286 +310,224 @@ class Orchestrator {
+      inputs: loopStep.inputs,
+    }
 
-      if (isFailure) {
-        this.updateContextAndOutput(
-          pathStepIdx + 1,
-          steps[stepToLoopIndex],
-          {
-            items: this.loopStepOutputs,
-            iterations: loopStepIndex,
-          },
-          {
-            status: AutomationErrors.FAILURE_CONDITION,
-            success: false,
-          }
-        )
-        shouldCleanup = false
-        break
-      }
+    const loopMaxIterations =
+      typeof loopStep.inputs.iterations === "string"
+        ? parseInt(loopStep.inputs.iterations)
+        : loopStep.inputs.iterations
+    const maxIterations = Math.min(
+      loopMaxIterations || env.AUTOMATION_MAX_ITERATIONS,
+      env.AUTOMATION_MAX_ITERATIONS
+    )
 
-      this.context.steps[pathStepIdx] = {
-        currentItem: this.getCurrentLoopItem(loopStep, loopStepIndex),
+    const items: Record<string, any>[] = []
+    let iterations = 0
+    let iterable: any[] = []
+    try {
+      iterable = getLoopIterable(loopStep)
+    } catch (err) {
+      return {
+        ...result,
+        outputs: {
+          success: false,
+          status: AutomationStepStatus.INCORRECT_TYPE,
+        },
       }
-
-      stepToLoopIndex = stepIdx + 1
-
-      await this.executeStep(steps[stepToLoopIndex], stepToLoopIndex)
-      iterationCount++
     }
 
-    if (shouldCleanup) {
-      let tempOutput =
-        iterations === 0
-          ? {
-              status: AutomationStepStatus.NO_ITERATIONS,
-              success: true,
-            }
-          : {
-              success: true,
-              items: this.loopStepOutputs,
-              iterations: iterationCount,
-            }
+    for (; iterations < iterable.length; iterations++) {
+      const currentItem = iterable[iterations]
 
-      if (reachedMaxIterations && iterations !== 0) {
-        tempOutput.status = AutomationStepStatus.MAX_ITERATIONS
+      if (iterations === maxIterations) {
+        return {
+          ...result,
+          outputs: {
+            success: false,
+            iterations,
+            items,
+            status: AutomationStepStatus.MAX_ITERATIONS,
+          },
+        }
       }
 
-      // Loop Step clean up
-      this.executionOutput.steps.splice(pathStepIdx, 0, {
-        id: steps[stepToLoopIndex].id,
-        stepId: steps[stepToLoopIndex].stepId,
-        outputs: tempOutput,
-        inputs: steps[stepToLoopIndex].inputs,
-      })
-
-      this.context.stepsById[steps[stepToLoopIndex].id] = tempOutput
-      const stepName = steps[stepToLoopIndex].name || steps[stepToLoopIndex].id
-      this.context.stepsByName[stepName] = tempOutput
-      this.context.steps[this.context.steps.length] = tempOutput
-      this.context.steps = this.context.steps.filter(
-        item => !item.hasOwnProperty.call(item, "currentItem")
-      )
+      if (matchesLoopFailureCondition(loopStep, currentItem)) {
+        return {
+          ...result,
+          outputs: {
```

Comment:
it would be nice to have a utility for setting a loop failing, just to reduce the visual load of this repeating, all that really matters here is the change in status, something like `return loopFailure(result, AutomationStepStatus.FAILURE_CONDITION)` would make this all look a lot clearer (to me at least)